README.md
# Flutter Custom Cards

Flutter Custom Cards help developers to create beautiful custom cards and 3D cards with flutter.

![Pub Badge](https://img.shields.io/pub/v/flutter_custom_cards?color=blueviolet&style=for-the-badge)
![License](https://img.shields.io/github/license/pktintali/flutter_custom_cards?style=for-the-badge)
![Dart Code](https://img.shields.io/github/languages/top/pktintali/flutter_custom_cards?color=success&style=for-the-badge)
![Repo Size](https://img.shields.io/github/repo-size/pktintali/flutter_custom_cards?color=ff69b4&style=for-the-badge)

## Demo Screenshot
<img src="https://raw.githubusercontent.com/pktintali/flutter_custom_cards/main/example/screenshots/demo.png" width="400">

## Getting Started

Add the package In your pubspec.yaml

```yaml
flutter_custom_cards: ^0.2.1
```

Import and use

```dart
import 'package:flutter_custom_cards/flutter_custom_card.dart'
```

## Basic Usages

To create a `CustomCard`
```dart
CustomCard(
  borderRadius: 10,
  color: Colors.yellow,
  hoverColor: Colors.indigo,
  onTap: () {},
  child: Text('Flutter'),
),
```

To Create a `Custom3DCard`

```dart
Custom3DCard(
  elevation: 4,
  child: SizedBox(
    height: 100,
    width: 100,
    child: Center(
      child: FlutterLogo(size: 65),
    ),
  ),
),
```
>For Full Uses See [Example](https://pub.dev/packages/flutter_custom_cards/example) Section
## Major Changes
On version 0.2 the following major changes were done on the release

- `TextCard`, `ImageCard` and `WidgetCard` have been merged into single card named `CustomCard`.
- Now new CustomCards supports onTap, hoverColor, height, width and more.
- `Custom3DCard` is introduced in this version that was not in the previous version.
- Now there are only two cards `CustomCard` and `Custom3DCard`

## CustomCard Options
The following options are available on `CustomCard`

| Property      | Type | Description |
| --------------| --------- |--------------|
| borderColor | Color       |Border color of card
| borderRadius   | double        |Radius of the card
| borderWidth   | double        | Border width of card
| child   | Widget        |child Widget of card
| childPadding   | double        | Padding for the child widget (default is 5)
| color   | Color        |Background `Color` for card
| elevation   | double        | Elevation for the card (default is 3)
| height   | double        | Height of the card (if null the height will be according to child widget)
| hoverColor   | Color        | Hover color of card (only visible if `onTap` is not null)
| key   | Key        | Key is an identifier for card
| onTap   | GestureTapCallback   |Signature for when a tap has occurred.
| shadowColor   | Color        | Shadow color of card
| splashColor   | Color        |Splash color of card (only visible if `onTap` is not null
| width   | double        |Width of the card (if null the width will be according to child widget)

## Custom3DCard Options
The following options are available on `Custom3DCard`

| Property      | Type | Description |
| --------------| --------- |--------------|
| borderOnForeground   | bool        |If false, the border will be painted behind the child (default true)
| child   | Widget        | Child widget of card
| color   | Color        | Color of card
| elevation   | double        |Elevation of card
| key   | Key        |Key is an identifier for card
| margin   | EdgeInsetsGeometry        |The empty space that surrounds the card
| semanticContainer   | bool        |Whether this widget represents a single semantic container (default true)
| shadowColor   | Color        |The color to paint the shadow below the card.
| shadowSpread   | double        | Spread distance of `shadow` below card (default is 10)
| shape   | ShapeBorder        | The shape of the card


## Upcoming
The following features will rollout soon
- `onTap` option on `Custom3DCard` as well
- `onHover` option in both of the cards
- `onLongPressed` and `onDoubleTap` option in both of the cards
- `style` property in `Custom3DCard` using this we can create different types of 3d cards.


## License
This project is licensed under the MIT license. See [LICENSE](https://github.com/pktintali/flutter_custom_cards/blob/main/LICENSE) for details.


<p align="center">
  <img alt = "Counter" src="https://profile-counter.glitch.me/flutter_custom_cards/count.svg"/>
</p>

CHANGELOG.md
# Changelog

## [0.2.1] - 21/11/2021

- Readme Docs Updated

## [0.2.0] - 20/11/2021

This is a big update (major changes)

The following **major changes** happened
- `TextCard`, `ImageCard` and `WidgetCard` have been merged into single card named `CustomCard`.
- Now new CustomCards supports onTap, hoverColor, height, width and more.
- `Custom3DCard` is introduced in this version that was not in the previous version.
- Now there are only two cards `CustomCard` and `Custom3DCard`

## [0.1.0] - 25/03/2021

- Migrated to `null-safety`

## [0.0.3] - 12/12/2020

- TextCard,ImageCard,WidgetCard Added

## [0.0.2] - 12/12/2020

- Some New Properties Added

## [0.0.1] - 11/12/2020

- Initial release, New Cards Coming Soon.

lib/src/TextCard.dart
import 'package:flutter/material.dart';

///[TextCard] that display text inside a card
class TextCard extends StatefulWidget {
  /// [String] that display text on screen
  final String text;

  ///Background [Color] for card
  final Color color;

  ///[Color] for the text
  final Color textColor;

  ///key
  final Key? key;

  ///[fontSize] of the [Text]
  final double fontSize;

  ///[shadow] color of the card
  final Color shadowColor;

  //padding for the child
  final double childPadding;

  ///[elevation] for card
  final double elevation;
  TextCard(
      {required this.text,
      this.elevation = 3,
      this.shadowColor = Colors.black,
      this.childPadding = 5,
      this.color = Colors.white,
      this.textColor = Colors.black,
      this.fontSize = 18.0,
      this.key});
  @override
  _CustomCardsState createState() => _CustomCardsState();
}

class _CustomCardsState extends State<TextCard> {
  @override
  Widget build(BuildContext context) {
    return Card(
      elevation: widget.elevation,
      key: widget.key,
      color: widget.color,
      shadowColor: widget.shadowColor,
      child: Padding(
        padding: EdgeInsets.all(widget.childPadding),
        child: Text(
          widget.text,
          style: TextStyle(
            color: widget.textColor,
            fontSize: widget.fontSize,
          ),
        ),
      ),
    );
  }
}


lib/src/ImageCard.dart
import 'package:flutter/material.dart';

class ImageCard extends StatefulWidget {
  ///Main [Image] for the card
  final Image image;

  ///Background [Color]
  final Color color;

  ///[shadow] color of the card
  final Color shadowColor;

  //key
  final Key? key;

  //padding for the child
  final double childPadding;

  ///[elevation] for card
  final double elevation;
  ImageCard(
      {required this.image,
      this.elevation = 3,
      this.shadowColor = Colors.black,
      this.childPadding = 5,
      this.color = Colors.white,
      this.key});
  @override
  _CustomCardsState createState() => _CustomCardsState();
}

class _CustomCardsState extends State<ImageCard> {
  @override
  Widget build(BuildContext context) {
    return Card(
      shadowColor: widget.shadowColor,
      elevation: widget.elevation,
      key: widget.key,
      color: widget.color,
      child: Padding(
        padding: EdgeInsets.all(widget.childPadding),
        child: widget.image,
      ),
    );
  }
}


lib/src/WidgetCard.dart
import 'package:flutter/material.dart';

class WidgetCard extends StatefulWidget {
  ///Main [Widget] inside the card
  final Widget widget;

  ///Background [Color] for card
  final Color color;

  //key
  final key;

  ///padding for the child widget
  final double widgetPadding;

  ///[elevation] for card
  final double elevation;
  WidgetCard(
      {required this.widget,
      this.elevation = 3,
      this.widgetPadding = 5,
      this.color = Colors.white,
      this.key});
  @override
  _CustomCardsState createState() => _CustomCardsState();
}

class _CustomCardsState extends State<WidgetCard> {
  @override
  Widget build(BuildContext context) {
    return Card(
      elevation: widget.elevation,
      key: widget.key,
      color: widget.color,
      child: Padding(
          padding: EdgeInsets.all(widget.widgetPadding), child: widget.widget),
    );
  }
}


lib/src/CustomCards.dart
import 'package:flutter/material.dart';

class CustomCards extends StatefulWidget {
  final String text;
  final Color color;
  final Color textColor;
  final key;
  final double fontSize;
  final double childPadding;
  final double elevation;
  CustomCards(
      {@required this.text,
      this.elevation=3,
      this.childPadding = 5,
      this.color = Colors.white,
      this.textColor = Colors.black,
      this.fontSize = 18.0,
      this.key});
  @override
  _CustomCardsState createState() => _CustomCardsState();
}

class _CustomCardsState extends State<CustomCards> {
  @override
  Widget build(BuildContext context) {
    return Card(
      elevation: widget.elevation,
      key: widget.key,
      color: widget.color,
      child: Padding(
        padding: EdgeInsets.all(widget.childPadding),
        child: Text(
          widget.text,
          style: TextStyle(
            color: widget.textColor,
            fontSize: widget.fontSize,
          ),
        ),
      ),
    );
  }
}


lib/src/custom_card.dart
import 'package:flutter/material.dart';

///[CustomCard] that you can use for quick mplementation of [Card]
///Many properties simplified
///Added more features to simple [Card] like you can click on [CustomCard]
///also support hoverColor and more
class CustomCard extends StatelessWidget {
  ///Background [Color] for card
  final Color color;

  ///key
  final Key? key;

  ///Creates an ink well.
  ///What should be happen when you click on [Card]
  final GestureTapCallback? onTap;

  ///shadow [Color] of the card
  final Color shadowColor;

  ///border color of card
  final Color? borderColor;

  ///hover [Color] for the [Card]
  ///onTap shoult not be null
  final Color? hoverColor;

  ///splash [Color] for click effect
  ///onTap shoult not be null
  final Color? splashColor;

  ///padding for the child
  final double childPadding;

  ///height of the [Card] if null then the height of [Card] will be according to child
  final double? height;

  ///height of the [Card] if null then the width of [Card] will be according to child
  final double? width;

  ///radius for the [Card]
  final double? borderRadius;

  ///The width of this side of the border, in logical pixels.
  final double? borderWidth;

  ///[elevation] for card
  final double elevation;

  ///child [Widget] of [Card]
  final Widget? child;

  CustomCard({
    this.elevation = 3,
    this.shadowColor = Colors.black,
    this.childPadding = 5,
    this.color = Colors.white,
    this.child,
    this.key,
    this.borderRadius,
    this.borderWidth,
    this.borderColor,
    this.hoverColor,
    this.splashColor,
    this.onTap,
    this.height,
    this.width,
  });
  @override
  Widget build(BuildContext context) {
    bool _isBorder = (borderColor != null || borderWidth != null);
    return Card(
      elevation: elevation,
      key: key,
      color: color,
      shadowColor: shadowColor,
      shape: RoundedRectangleBorder(
        borderRadius: BorderRadius.circular(borderRadius ?? 0),
        side: _isBorder
            ? BorderSide(
                color: borderColor ?? Colors.black,
                width: borderWidth ?? 1,
                style: BorderStyle.solid,
              )
            : BorderSide.none,
      ),
      child: InkWell(
        borderRadius: BorderRadius.circular(borderRadius ?? 0),
        hoverColor: hoverColor,
        splashColor: splashColor,
        onTap: onTap,
        child: SizedBox(
          height: height,
          width: width,
          child: Padding(
            padding: EdgeInsets.all(childPadding),
            child: child,
          ),
        ),
      ),
    );
  }
}


lib/src/custom_3d_card.dart
import 'package:flutter/material.dart';

class Custom3DCard extends StatelessWidget {
  const Custom3DCard({
    Key? key,
    this.color,
    this.shadowColor,
    this.elevation,
    this.shape,
    this.borderOnForeground = true,
    this.margin,
    this.clipBehavior,
    this.child,
    this.shadowSpread = 10,
    this.semanticContainer = true,
  })  : assert(elevation == null || elevation >= 0.0),
        super(key: key);

  /// The card's background color.
  ///
  /// Defines the card's [Material.color].
  ///
  /// If this property is null then [CardTheme.color] of [ThemeData.cardTheme]
  /// is used. If that's null then [ThemeData.cardColor] is used.
  final Color? color;

  /// The color to paint the shadow below the card.
  ///
  /// If null then the ambient [CardTheme]'s shadowColor is used.
  /// If that's null too, then the overall theme's [ThemeData.shadowColor]
  /// (default black) is used.
  final Color? shadowColor;

  /// the total spread of outer shadow of [Card]
  /// default to
  final double shadowSpread;

  /// The z-coordinate at which to place this card. This controls the size of
  /// the shadow below the card.
  ///
  /// Defines the card's [Material.elevation].
  ///
  /// If this property is null then [CardTheme.elevation] of
  /// [ThemeData.cardTheme] is used. If that's null, the default value is 1.0.
  final double? elevation;

  /// The shape of the card's [Material].
  ///
  /// Defines the card's [Material.shape].
  ///
  /// If this property is null then [CardTheme.shape] of [ThemeData.cardTheme]
  /// is used. If that's null then the shape will be a [RoundedRectangleBorder]
  /// with a circular corner radius of 4.0.
  final ShapeBorder? shape;

  /// Whether to paint the [shape] border in front of the [child].
  ///
  /// The default value is true.
  /// If false, the border will be painted behind the [child].
  final bool borderOnForeground;

  /// {@macro flutter.material.Material.clipBehavior}
  ///
  /// If this property is null then [CardTheme.clipBehavior] of
  /// [ThemeData.cardTheme] is used. If that's null then the behavior will be [Clip.none].
  final Clip? clipBehavior;

  /// The empty space that surrounds the card.
  ///
  /// Defines the card's outer [Container.margin].
  ///
  /// If this property is null then [CardTheme.margin] of
  /// [ThemeData.cardTheme] is used. If that's null, the default margin is 4.0
  /// logical pixels on all sides: `EdgeInsets.all(4.0)`.
  final EdgeInsetsGeometry? margin;

  /// Whether this widget represents a single semantic container, or if false
  /// a collection of individual semantic nodes.
  ///
  /// Defaults to true.
  ///
  /// Setting this flag to true will attempt to merge all child semantics into
  /// this node. Setting this flag to false will force all child semantic nodes
  /// to be explicit.
  ///
  /// This flag should be false if the card contains multiple different types
  /// of content.
  final bool semanticContainer;

  /// The widget below this widget in the tree.
  ///
  /// {@macro flutter.widgets.ProxyWidget.child}
  final Widget? child;

  @override
  Widget build(BuildContext context) {
    return DecoratedBox(
      child: Card(
        color: color,
        elevation: 0,
        child: Padding(
          padding: const EdgeInsets.all(8.0),
          child: DecoratedBox(
            child: Card(
              child: child,
              borderOnForeground: borderOnForeground,
              clipBehavior: clipBehavior,
              color: color,
              elevation: elevation,
              margin: margin,
              semanticContainer: semanticContainer,
              shadowColor: shadowColor,
              shape: shape,
              key: key,
            ),
            decoration: BoxDecoration(
              boxShadow: [
                BoxShadow(
                  color: shadowColor ?? Colors.black12,
                  blurRadius: 15.0,
                  spreadRadius: 2.5,
                ),
              ],
            ),
          ),
        ),
      ),
      decoration: BoxDecoration(
        boxShadow: [
          BoxShadow(
            color: shadowColor ?? Colors.black12,
            blurRadius: shadowSpread,
            spreadRadius: 1.0,
          ),
        ],
      ),
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flutter_custom_cards/flutter_custom_cards.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'CustomCards',
      debugShowCheckedModeBanner: false,
      home: CustomCardsExample(),
    );
  }
}

class CustomCardsExample extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        centerTitle: true,
        title: Text('Custom Cards Example'),
      ),
      body: SingleChildScrollView(
        child: Padding(
          padding: const EdgeInsets.all(8.0),
          child: Column(
            children: [
              Text('CustomCards'),
              Center(
                child: Wrap(
                  //mainAxisAlignment: MainAxisAlignment.center,
                  children: [
                    CustomCard(
                      height: 50,
                      child: FlutterLogo(
                        style: FlutterLogoStyle.horizontal,
                        size: 90,
                      ),
                    ),
                    CustomCard(
                      height: 50,
                      width: 100,
                      elevation: 6,
                      childPadding: 10,
                      color: Colors.green,
                      onTap: () {},
                      child: Center(
                        child: Text(
                          'Flutter',
                          style: TextStyle(
                            fontSize: 21,
                            color: Colors.white,
                          ),
                        ),
                      ),
                    ),
                    CustomCard(
                      height: 50,
                      width: 100,
                      borderRadius: 10,
                      color: Colors.red,
                      hoverColor: Colors.indigo,
                      splashColor: Colors.white,
                      onTap: () {},
                      child: Center(
                        child: Text(
                          'Flutter',
                          style: TextStyle(
                            fontSize: 20,
                            color: Colors.white,
                          ),
                        ),
                      ),
                    ),
                  ],
                ),
              ),
              SizedBox(height: 8),
              Wrap(
                //mainAxisAlignment: MainAxisAlignment.center,
                children: [
                  CustomCard(
                    height: 100,
                    width: 100,
                    elevation: 5,
                  ),
                  CustomCard(
                    height: 100,
                    width: 100,
                    elevation: 5,
                    borderRadius: 50,
                    color: Colors.green,
                    hoverColor: Colors.yellow,
                    onTap: () {},
                  ),
                  CustomCard(
                    height: 100,
                    width: 100,
                    elevation: 5,
                    color: Colors.blue,
                    borderColor: Colors.white,
                    borderWidth: 2,
                  ),
                ],
              ),
              SizedBox(height: 25),
              Text('Custom3DCards'),
              Wrap(
                //mainAxisAlignment: MainAxisAlignment.center,
                children: [
                  Padding(
                    padding: const EdgeInsets.all(15.0),
                    child: Custom3DCard(
                      elevation: 4,
                      child: SizedBox(
                        height: 100,
                        width: 100,
                        child: Center(
                          child: FlutterLogo(size: 65),
                        ),
                      ),
                    ),
                  ),
                  Padding(
                    padding: const EdgeInsets.all(15.0),
                    child: Custom3DCard(
                      elevation: 10,
                      shadowSpread: 5,
                      shadowColor: Colors.brown.shade400,
                      child: SizedBox(
                        height: 100,
                        width: 100,
                        child: Center(
                          child: FlutterLogo(
                            size: 65,
                            style: FlutterLogoStyle.stacked,
                          ),
                        ),
                      ),
                    ),
                  ),
                ],
              ),
              SizedBox(height: 30),
            ],
          ),
        ),
      ),
    );
  }
}
